File: 31-03-2025_Put_Marbles_in_Bags.py
class Solution(object):
    def putMarbles(self, weights, k):
        if k == 1:
            return 0  
        n = len(weights)
    
    # Create pairwise adjacent sums
        adjacent_sums = [weights[i] + weights[i + 1] for i in range(n - 1)]
    
    # Sort the adjacent sums
        adjacent_sums.sort()
    
    # Calculate min and max scores
        min_score = sum(adjacent_sums[:k-1])
        max_score = sum(adjacent_sums[-(k-1):])
    
        return max_score - min_score

File: 31-03-2025_Put_Marbles_in_Bags.java
class Solution {
    //public long putMarbles(int[] weights, int k) {
        
    //}
    public  long putMarbles(int[] weights, int k) {
        if (k == 1) {
            return 0;  // Only one partition, so no difference.
        }
        
        int n = weights.length;
        long[] pairs = new long[n - 1];
        
        // Create pairwise adjacent sums
        for (int i = 0; i < n - 1; i++) {
            pairs[i] = (long) weights[i] + weights[i + 1];
        }
        
        // Sort the pairs
        Arrays.sort(pairs);
        
        // Calculate min and max scores
        long minScore = 0;
        long maxScore = 0;
        
        for (int i = 0; i < k - 1; i++) {
            minScore += pairs[i];
            maxScore += pairs[n - 2 - i];
        }
        
        return maxScore - minScore;
    }
}
